### PR TITLE
Add CRD and type definition of new resource TidbMonitor

### PR DESCRIPTION
**Pull request by aylei**
Signed-off-by: Aylei <rayingecho@gmail.com>

#1121 
#1261 

# release note
```
New resource TidbMonitor for managing TiDB cluster monitoring components.
```

@Yisaer @cofyc @tennix @weekface @qiffang PTAL

I've separate the commits to ease the review process:

- commit https://github.com/pingcap/tidb-operator/pull/1331/commits/1c1d848058d16bbf7cc77f2cf74eaae97f3a717a  is the type definition and related changes
- commit https://github.com/pingcap/tidb-operator/pull/1331/commits/e1bff79626d8588a000fb1daf1b9adbc490d83c0 contains the generated code

### CODE REVIEW

**Comment by aylei** [comment]
@LinuxGit @qiffang Please check the API to see if it feasible for you to integrate.

File: pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+	// +k8s:openapi-gen=false
+	// Most recently observed status of the TidbMonitor
+	Status TidbMonitorStatus `json:"status"`
+}
+
+// +k8s:openapi-gen=true
+// TidbMonitor spec encode the desired state of tidb monitoring component
+type TidbMonitorSpec struct {
+	Clusters        []TidbClusterRef  `json:"clusters"`
```
Comment (Yisaer) [inline pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go]: Does the `Clusters` means one `TidbMonitor` could monitor serveral `TidbCluster`?

Comment (aylei) [reply pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go]: Yes

File: pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+	Prometheus      PrometheusSpec    `json:"prometheus"`
+	Grafana         GrafanaSpec       `json:"grafana"`
```
Comment (weekface) [inline pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go]: grafana may be `optional `

Comment (aylei) [reply pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go]: good catch

File: pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+}
+
+// PrometheusSpec is the desired state of prometheus
+type PrometheusSpec struct {
+	MonitorContainer `json:",inline"`
+
+	LogLevel    string      `json:"logLevel,omitempty"`
+	ReserveDays int         `json:"reserveDays,omitempty"`
+	Service     ServiceSpec `json:"service,omitempty"`
```
Comment (LinuxGit) [inline pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go]: TEP relies on the alertmanagerURL field https://github.com/pingcap/tidb-operator/blob/master/charts/tidb-cluster/values.yaml#L517.

File: pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+type ReloaderSpec struct {
+	MonitorContainer `json:",inline"`
+	Service          ServiceSpec `json:"service,omitempty"`
+}
+
+// InitializerSpec is the desired state of initializer
+type InitializerSpec struct {
+	MonitorContainer `json:",inline"`
+	// +optional
+	Envs map[string]string `json:"envs,omitempty"`
```
Comment (LinuxGit) [inline pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go]: We could set K8S_PROMETHEUS_URL via Envs? Grafana nodesInfo dashboard relies on this field.

Comment (aylei) [reply pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go]: Yes we could

File: pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go (hunk 3 above)
Comment (aylei) [reply pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go]: Good catch, I will add it

File: pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+// +k8s:openapi-gen=true
+// TidbMonitor spec encode the desired state of tidb monitoring component
+type TidbMonitorSpec struct {
+	Clusters []TidbClusterRef `json:"clusters"`
+
+	Prometheus PrometheusSpec `json:"prometheus"`
+	// +optional
+	Grafana     *GrafanaSpec    `json:"grafana"`
```
Comment (tennix) [inline pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go]: Add `omitempty` to not display this field when `Grafana` is not available.

File: pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+	// +optional
+	Annotations map[string]string `json:"annotations,omitempty"`
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+
+	// kubePrometheusURL is where tidb-monitoring get the  common metrics of kube-prometheus.
+	// Ref: https://github.com/coreos/kube-prometheus
+	// +optional
+	KubePrometheusURL string `json:"kubePrometheusURL,omitempty"`
```
Comment (tennix) [inline pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go]: Should we define these optional fields as pointers?

Comment (aylei) [reply pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go]: It depends on whether empty string (`""`) should be considered as "unset" for this field (typically, field cannot be unset when we merge changes to existing object, e.g. server-side apply), THIS one should is better to be a pointer I guess, I will check whether other `string` fields should be changed to `*string`

File: pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go (hunk 5 above)
Comment (aylei) [reply pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go]: good catch

**Comment by aylei** [comment]
@tennix addressed, PTAL

@LinuxGit I've addressed some comments, PTAL again

**Pull request merged**